sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -741,6 +889,11 @@ kernel_pid_t gcoap_init(void)
     if (_pid != KERNEL_PID_UNDEF) {
         return -EEXIST;
     }
+
+#ifdef CONFIG_GCOAP_ENABLE_DTLS
+    _main_thread = thread_getpid();
```

Comment:
No, every thread can do CoAP communication. Now that I've looked at it again: This part should be moved to `_sock_send` and is not limited to the main_thread. 

--
To explain the idea behind this mechanic: Gcoap does only store the data that is to be sent on client side and when it is a confirmable message. It stores it in a memo which can be refound when the answer to the request arrives. 

If we want to send data and a session to the endpoint has not been established before we need to do a handshake first. Ideally you would initiate the handshake and when the handshake process is completed we send the data. If we want the handshake process to be asynchronous on gcoap side too, we need to store the data separate because, currently, we do not hold the data until the handshake is completed. 

Storing that data would cost `DTLS_HANDSHAKE_MAX` * `CONFIG_GCOAP_PDU_BUF_SIZE` (default 128 bytes) and we would need to attach this data somehow to the session (doable via session management I guess). `DTLS_HANDSHAKE_MAX` is the maximum of concurrent handshakes. 

That's a lot of additional memory and the reason why I decided to do it with blocking handshakes. When a session needs to be established we save the thread doing the handshake initiating and then wait till the process completes or timeouts.

When the successful handshake "arrives" in the DTLS event handler we inform the blocked thread that the handshake has been successful and completed -> `_sock_send` blocks until handshake is completed and can send data. We do not have to store additional data this way.

But it comes with a price - the gcoap server cannot do handshakes on its own this way. At the moment this is not needed. But in the meantime I have seen that it could(!) be necessary for observable resources. I have not yet fully understood a statement in the RFC standard regarding that.